lib/internal/modules/esm/resolve.js

```diff
@@ -812,22 +813,40 @@ function shouldBeTreatedAsRelativeOrAbsolutePath(specifier) {
  * @returns {url: URL, format?: string}
  */
 function moduleResolve(specifier, base, conditions, preserveSymlinks) {
-  const isRemote = base.protocol === 'http:' ||
-    base.protocol === 'https:';
+  const protocol = typeof base === 'string' ?
+    StringPrototypeSlice(base, 0, StringPrototypeIndexOf(base, ':') + 1) :
+    base.protocol;
+  const isData = protocol === 'data:';
+  const isRemote =
+    isData ||
+    protocol === 'http:' ||
+    protocol === 'https:';
   // Order swapped from spec for minor perf gain.
   // Ok since relative URLs cannot parse as URLs.
   let resolved;
   if (shouldBeTreatedAsRelativeOrAbsolutePath(specifier)) {
+    if (isData) {
+      try {
+        new URL(specifier, base); // This is going to throw.
```

Comment:
Can we use `URLCanParse`?

Reply:
Would `URLCanParse` throw?

Reply:
No, it won't

Reply:
Then it wouldn't work here for what I'm trying to do, right?

Reply:
This would be a lot faster:

```
if (!URL.canParse(specifier, base)) {
  const error = ERR_INVALID_URL(specifier);
  error.cause = new ERR_UNSUPPORTED_RESOLVE_REQUEST(specifier);
  throw error;
}
```

Reply:
PS: `URL.canParse` has V8 Fast API optimization

Reply:
If you expect it to throw, then we can just eliminate URL.canParse call, and just construct the error in here. Right?

Reply:
Yeah the `URLCanParse` is really not necessary. But I'm not sure I want to create the error here, that would seem a bit out of place, and make the code harder to maintain if we ever change what `new URL` would throw. Also, it's  the error path, is it worth optimizing?

Reply:
You might be right about "is it worth optimizing" part, but since `ERR_INVALID_URL` is defined inside `internals/errors.js` and can be technically required by any file in the project, I don't think it is an out of place to construct it here. I think `new URL()` is just unnecessary in this part, since we know that it will throw. But I'll resolve this and keep it up to you.

Reply:
Well you are probably at a better place than me to decide if it would add maintenance burden to URL, so I'm fine going with you suggestion if you stand by it.

Reply:
Let's go for it!